hail/python/hailtop/aiocloud/aioterra/azure/fs.py

```diff
@@ -0,0 +1,120 @@
+import asyncio
+import os
+from typing import AsyncContextManager, AsyncIterator, Dict, Set, Tuple, Optional
+
+from hailtop.aiocloud.aioazure import AzureAsyncFS, AzureAsyncFSURL
+from hailtop.aiotools.fs import AsyncFS, ReadableStream, WritableStream, MultiPartCreate, FileStatus, FileListEntry
+from hailtop.utils import time_msecs
+
+
+from .client import TerraClient
+
+
+WORKSPACE_STORAGE_CONTAINER_ID = os.environ.get('WORKSPACE_STORAGE_CONTAINER_ID')
+WORKSPACE_STORAGE_CONTAINER_URL = os.environ.get('WORKSPACE_STORAGE_CONTAINER_URL')
+
+
+class TerraAzureAsyncFS(AsyncFS):
+
+    def __init__(self, **azure_kwargs):
+        self._terra_client = TerraClient()
+        self._azure_fs = AzureAsyncFS(**azure_kwargs)
+        self._sas_token_cache: Dict[str, Tuple[str, int]] = {}
+
+    @property
+    def schemes(self) -> Set[str]:
+        return {'https'}
+
+    @staticmethod
+    def configured() -> bool:
+        return WORKSPACE_STORAGE_CONTAINER_ID is not None and WORKSPACE_STORAGE_CONTAINER_URL is not None
+
+    @staticmethod
+    def valid_url(url: str) -> bool:
+        return AzureAsyncFS.valid_url(url)
+
+    def parse_url(self, url: str) -> AzureAsyncFSURL:
+        return AzureAsyncFS.parse_url(url)
+
+    async def _to_azure_url(self, url: str) -> str:
+        if url in self._sas_token_cache:
```

Comment:
can we use the walrus operator now?

Reply:
Unfortunately I don't believe that tuple deconstruction works with the walrus operator, so I'm not sure I see much value.